book/src/dev/rfcs/xxxx-block-subsidy.md

```diff
@@ -0,0 +1,313 @@
+
+```
+pub enum SubsidyError {
+    #[error("not a coinbase transaction")]
+    NoCoinbase,
+
+    #[error("founders reward amount not found")]
+    FoundersRewardAmountNotFound,
+    
+    #[error("founders reward address not found")]
+    FoundersRewardAddressNotFound,
+    
+    #[error("founding stream amount not found")]
+    FundingStreamAmountNotFound,
+    
+    #[error("founding stream address not found")]
+    FundingStreamAddressNotFound,
+    
+    #[error("the sum of outputs is greater than calculated subsidies and transaction fees")]
+    MinerSubsidyRuleBroken,
+    
+    #[error("invalid shielded descriptions found")]
+    ShieldedDescriptionsInvalid,
+
+    #[error("broken rule in shielded transaction inside coinbase")]
+    ShieldedRuleBroken,
+}
+```
+## Implementation Plan
+
+1. Founders Reward Amounts
+2. Founders Reward Addresses
+3. Funding Streams Amounts (defer Shielded Coinbase)
+4. Funding Streams Addresses 
+5. Miner Subsidy Amounts (defer Shielded Coinbase)
+6. Shielded Coinbase for Funding Streams and Miner Subsidy
+
+Each stage should have code, unit tests, block test vector tests, and property tests, before moving on to the next stage. (A stage can be implemented in multiple PRs.)
+
+## Test Plan
+
+For each network(Mainnet, Testnet), calculation of subsidy amounts need a `Height` as input and will output different amounts according to it.
+
+- Test all subsidy amount functions(`block_subsidy()`, `miner_subsidy()`, `founders_reward()` and `funding_stream()`) against all network upgrade events of each network and make sure they return the expected amounts according to known outputs.
+
+For each network, the address of the reward receiver on each block will depend on the `Height`.
+
+- Test all subsidy address functions(`founders_reward_address()` and `funding_stream_address()`) against all network upgrade events of each network and make sure they return the expected addresses.
+
+Validation tests will test the consensus rules using real blocks from `zebra-test` crate. For both networks, blocks for all network upgrades were added to the crate in [#1096](https://github.com/ZcashFoundation/zebra/pull/1096). Blocks containing shielded coinbase were also introduced at [#1116](https://github.com/ZcashFoundation/zebra/pull/1116)
+
+- Test validation functions(`subsidy_is_valid()`) against all the blocks zebra haves available in the test vectors collection for both networks(`zebra_test::vectors::MAINNET_BLOCKS`  and `zebra_test::vectors::TESTNET_BLOCKS`), all blocks should pass validation.
+
+The validation errors at `SubsidyError` must be tested at least once.
+
+- Create tests to trigger each error from `SubsidyError`.
+
+- Create a property test for each consensus rule, which makes sure that the rule is satisfied by arbitrary (randomised) blocks and transactions
+  - Create or update `Arbitrary` trait implementations for blocks and transactions, as needed
```

Comment:
I'm not sure how useful these tests will be.

Let's work with the other developers to come up with a list of block subsidy invariants we could test using proptests.